src/game.cpp

```diff
@@ -3365,25 +3365,49 @@ bool game::save_achievements()
         std::replace_copy_if( std::begin( u.name ), std::begin( u.name ) + truncated_name_len,
                               std::ostream_iterator<char>( achievement_file_path ),
         [&]( const char c ) {
-            return !std::isgraph( c, locale );
+            return !std::isgraph( c, locale ) || !is_char_allowed( c );
         }, '_' );
     } else {
-        achievement_file_path << u.name;
+        std::replace_copy_if( std::begin( u.name ), std::begin( u.name ) + truncated_name_len,
+                              std::ostream_iterator<char>( achievement_file_path ),
+        [&]( const char c ) {
+            return !is_char_allowed( c );
+        }, '_' );
     }
 
     // Add a ~ if the player name was actually truncated.
     achievement_file_path << ( ( truncated_name_len != name_len ) ? "~-" : "-" );
     const int character_id = get_player_character().getID().get_value();
+
+    // Add a timestamp for uniqueness.
```

Comment:
The problem with this approach is that a different achievement file is saved each time you save a character so this will create multiple copies of achievement files of a single character.

Reply:
hmm, indeed, I just forget the save is repeatable

Reply:
You can take world creation datetime. It is probably unique

Reply:
> You can take world creation datetime. It is probably unique

There is a situation, in same world, use character with special char in same position. Dose the creation time record when character be created?

Reply:
Sorry, probably not

Reply:
I have an idea, just make a map, like which I said above.
`a|a|a|a` convert to `a_a_a_a-_73_73_73_` the position of `_ ` in `a_a_a_a`  store the postion of special characters, then `_ ` in `_73_73_73_` store  the postion of valid characters, and the number store char in ascii. then the name will be unique and don't need many operation.

Reply:
The character id added in line 3379 is probably unique in a world, so with the world creation time it should be unique enough for our purpose.